### PR TITLE
MODINVSTOR-1069: Added SuDoc call number implementation

### PR DESCRIPTION
**Pull request by hasanepam**
## Purpose

Add new Call number implementation for SuDoc call numbers

## Approach

Added new class called SuDocCallNumber with unit test

## Learning
[MODINVSTOR-1069](https://issues.folio.org/browse/MODINVSTOR-1069)

### CODE REVIEW

**Review by mukhiddin-yusuf — APPROVED** [review]

File: src/test/java/org/folio/services/CallNumberUtilsTest.java [hunk 1]
```diff
@@ -75,4 +80,69 @@ void inputForShelvingNumber(
...
+    "T22.19/2:V88/retest/989,T22.19/2:V88/retest,T22.19/2:P94/2,"
+      + "T 222 219 12  !P 294 12 T 222 219 12  !V 288  !RETEST T 222 219 12  !V 288  !RETEST !3989",
+    "T22.19/2:V88/989/student/spanish,T22.19/2:V88/test/989,T22.19/2:V88/989/student/militia,"
+      + "T 222 219 12  !V 288  !TEST !3989 T 222 219 12  !V 288 !3989  !STUDENT  !MILITIA "
+      + "T 222 219 12  !V 288 !3989  !STUDENT  !SPANISH",
+    "T22.19/2:V88/2/989,T22.19:M54,T22.19:M54/990,T 222 219  !M 254 T 222 219  !M 254 !3990 "
+      + "T 222 219 12  !V 288 12 !3989",
+    "C 55.309/2-8,C 55.309/2,C 55.309/2-2,C 255 !3309 12 C 255 !3309 12 12 C 255 !3309 12 18",
+    "D 3.186/3,C 55.309/2-10,D 3.186,C 255 !3309 12 210 D 13 !3186 D 13 !3186 13",
+    "EP 1.23:91-44,D 3.186/7-3,EP 1.23:A 62 A 1.35,D 13 !3186 17 13 EP 11 223  !A 262  !A 11 235 EP 11 223 291 244"
```
[the remark this answers is not in the document]
Comment (hasanepam) [reply src/test/java/org/folio/services/CallNumberUtilsTest.java:96]: Got it, let me change it as you are not the first person who told this), but thank you!

File: src/main/java/org/folio/services/SuDocCallNumber.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
+package org.folio.services;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.apache.commons.lang.StringUtils;
+import org.marc4j.callnum.AbstractCallNumber;
+import org.marc4j.callnum.Utils;
+
+public class SuDocCallNumber extends AbstractCallNumber {
+  public static final String STEM_PATTERN = "^([A-Za-z]+\\s*)(\\d+)(\\.[A-Za-z]*\\d*)(/*[A-Za-z]*\\d*-*\\d*)";
```
[the remark this answers is not in the document]
Comment (hasanepam) [reply src/main/java/org/folio/services/SuDocCallNumber.java]: Replaced

**Review by viacheslavkol — CHANGES_REQUESTED** [review]

File: src/main/java/org/folio/services/CallNumberUtils.java [hunk 3]
```diff
@@ -1,21 +1,40 @@
...
+      return function.apply(callNumber);
+    }
 
-  public static Optional<String> getShelfKeyFromCallNumber(String callNumber) {
     return Optional.ofNullable(callNumber)
-      .flatMap(cn -> getValidShelfKey(new NlmCallNumber(cn))
-        .or(() -> getValidShelfKey(new LCCallNumber(cn)))
-        .or(() -> getValidShelfKey(new DeweyCallNumber(cn))))
-      .or(() -> Optional.ofNullable(callNumber))
       .map(String::trim);
```
[the remark this answers is not in the document]
Comment (hasanepam) [reply src/main/java/org/folio/services/CallNumberUtils.java:39]: done!

File: src/main/java/org/folio/rest/support/EffectiveCallNumberComponentsUtil.java [hunk 4]
```diff
@@ -24,6 +24,7 @@ public static Item calculateAndSetEffectiveShelvingOrder(Item item) {
     if (isNotBlank(item.getEffectiveCallNumberComponents().getCallNumber())) {
       Optional<String> shelfKey
         = CallNumberUtils.getShelfKeyFromCallNumber(
+          item.getItemLevelCallNumberTypeId(),
```
[the remark this answers is not in the document]
Comment (hasanepam) [reply src/main/java/org/folio/rest/support/EffectiveCallNumberComponentsUtil.java]: Understood, done!

**Review by viacheslavkol — APPROVED** [review]

**Pull request merged**